vdirsyncer/storage/ldap.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import ldap3
+import vobject
+
+from .base import Item, Storage
+from .. import exceptions
+
+ldap_logger = logging.getLogger(__name__)
+
+
+class LDAPStorage(Storage):
+    '''
+    :param url: LDAP URL
+    :param search_base: search base
+    :param bind: bind dn
+    :param password: bind password
+    :param filter: filter
+    '''
+    storage_name = 'ldap'
+    fileext = '.vcf'
+    item_mimetype = 'text/vcard'
+
+    def __init__(self, url='ldap://localhost', search_base=None, bind=None,
+                 password=None,
+                 filter='(&(objectCategory=person)(objectClass=user)'
+                        '(sn=*)(givenName=*))',
+                 _conn=None, **kwargs):
+        super(LDAPStorage, self).__init__(**kwargs)
+        self.search_base = search_base
+        self.filter = filter
+        self.conn = _conn
+        if self.conn is None:
+            server = ldap3.Server(url, get_info=ldap3.DSA)
+            if bind:
+                self.conn = ldap3.Connection(server, user=bind,
+                                             password=password)
+            else:
+                self.conn = ldap3.Connection(server)
+            self.conn.bind()
+            self.conn.start_tls()
+
+        ldap_logger.debug('Connected to: {}'.format(self.conn))
+
+        if self.search_base is None:
+            # Fallback to default root entry
+            self.search_base = server.info.naming_contexts[0]
+
+    def list(self):
+        '''
+        :returns: list of (href, etag)
+        '''
+        ldap_logger.debug('Search on {self.search_base} with filter'
+                          '{self.filter}'.format(self=self))
+        self.conn.search(self.search_base, self.filter,
+                         attributes=["whenChanged"])
+        for entry in self.conn.entries:
+            ldap_logger.debug('Found {}'.format(entry.entry_get_dn()))
+            href = entry.entry_get_dn()
+            if getattr(entry, 'whenChanged'):
+                etag = str(entry.whenChanged)
+            else:
+                item = self.get(href)
+                etag = item.hash
+
+            yield href, etag
+
+    def get(self, href):
+        self.conn.search(href, self.filter,
+                         attributes=["whenChanged", "cn", "sn", "givenName",
+                                     "displayName", "telephoneNumber",
+                                     "mobile", "facsimileTelephoneNumber",
+                                     "mail", "title"])
+
+        if not self.conn.entries[0]:
+            raise exceptions.NotFoundError(href)
+
+        entry = self.conn.entries[0]
+        etag = str(entry.whenChanged)
+
+        vcard = vobject.vCard()
+        vo = vcard.add('fn')
+        vo.value = str(entry.cn)
+        vo = vcard.add('n')
+        vo.value = vobject.vcard.Name(family=str(entry.sn),
+                                      given=str(entry.givenName))
+        if getattr(entry, 'telephoneNumber', None):
```

Comment:
This conversion from entry to item could be its own function (outside of storage class)